sandbox/apparmor/notify/listener/listener.go

```diff
@@ -0,0 +1,426 @@
+
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil/epoll"
+	"github.com/snapcore/snapd/sandbox/apparmor/notify"
+)
+
+var (
+	// ErrAlreadyRun indicates that the Run() method has already been run.
+	// Each listener must only be run once, so that spawned goroutines can
+	// be safely tracked and terminated.
+	ErrAlreadyRun = errors.New("listener has already been run")
+
+	// ErrAlreadyClosed indicates that the listener has been closed
+	ErrAlreadyClosed = errors.New("listener has already been closed")
+
+	// ErrAlreadyReplied indicates that the request has already received a reply
+	ErrAlreadyReplied = errors.New("request has already received a reply")
+
+	// ErrNotSupported indicates that the kernel does not support apparmor prompting
+	ErrNotSupported = errors.New("kernel does not support apparmor notifications")
+
+	osOpen      = os.Open
+	notifyIoctl = notify.Ioctl
+)
+
+// Request is a high-level representation of an apparmor prompting message.
+//
+// Each request must be replied to by writing a boolean to the YesNo channel.
+type Request struct {
+	// pid is the identifier of the process which triggered the request.
+	pid uint32
+	// label is the apparmor label on the process which triggered the request.
+	label string
+	// subjectUID is the UID of the subject which triggered the request.
+	subjectUID uint32
+
+	// path is the path of the file, as seen by the process triggering the request.
+	path string
+	// permission is the opaque permission that is being requested.
+	permission interface{}
+	// replyChan is a channel for writing the response.
+	replyChan chan interface{}
+	// replied indicates whether a reply has already been sent for this request.
+	replied uint32
+}
+
+func newRequest(msg *notify.MsgNotificationFile) (*Request, error) {
+	var perm interface{}
+	if msg.Class == notify.AA_CLASS_FILE {
+		_, missingPerms, err := msg.DecodeFilePermissions()
+		if err != nil {
+			return nil, err
+		}
+		perm = missingPerms
+	}
+	return &Request{
+		pid:        msg.Pid,
+		label:      msg.Label,
+		path:       msg.Name,
+		subjectUID: msg.SUID,
+
+		permission: perm,
+
+		replyChan: make(chan interface{}, 1),
+	}, nil
+}
+
+// PID returns the PID of the process which triggered the request.
+func (r *Request) PID() uint32 {
+	return r.pid
+}
+
+// Label returns the apparmor label on the process which triggered the request.
+func (r *Request) Label() string {
+	return r.label
+}
+
+// SubjectUID returns the UID of the subject which triggered the request.
+func (r *Request) SubjectUID() uint32 {
+	return r.subjectUID
+}
+
+// Path is the path of the file, as seen by the process which triggered the request.
+func (r *Request) Path() string {
+	return r.path
+}
+
+// Permission returns the opaque permission that is being requested.
+func (r *Request) Permission() interface{} {
+	return r.permission
+}
+
+// Reply sends the given response back to the kernel.
+func (r *Request) Reply(response interface{}) error {
```

Comment:
I don't think the shared `Reply` method should be aware of particular mediation classes or be responsible for validating them, since the reply must match the original mclass which produced the `Request`, and that information is not available in the `Request` (although it likely will need to be in the future, so the clients know how to interpret the requested permissions).

The goroutine waiting on the other end of the reply channel is in a mclass-specific function, so I think it's probably best if each of those functions is responsible for validating both the type of the response and its contents.

Reply:
this sounds like a recipe for subtle undetected bugs? can't we provide the Request with enough information to do the right thing?

Reply:
Yes, I'll include the mediation class in the request struct.